crates/rattler_conda_types/src/conda_lock/builder.rs

```diff
@@ -0,0 +1,209 @@
+    /// Metadata dealing with the time lockfile was created
+    pub time_metadata: Option<TimeMeta>,
+    /// Metadata dealing with the git repo the lockfile was created in and the user that created it
+    pub git_metadata: Option<GitMeta>,
+
+    /// Keep track of locked packages per platform
+    pub locked_packages: HashMap<Platform, LockedPackages>,
+
+    /// MatchSpecs input
+    /// This is only used to calculate the content_hash
+    /// for the lock file
+    pub input_specs: Vec<MatchSpec>,
+}
+
+impl LockFileBuilder {
+    pub fn new(
+        channels: impl IntoIterator<Item = Channel>,
+        platforms: impl IntoIterator<Item = Platform>,
+        input_spec: impl IntoIterator<Item = MatchSpec>,
+    ) -> Self {
+        Self {
+            channels: channels
+                .into_iter()
+                .map(|into_channel| into_channel.into())
+                .collect(),
+            platforms: platforms.into_iter().collect(),
+            input_specs: input_spec.into_iter().collect(),
+            ..Default::default()
+        }
+    }
+
+    /// Add locked packages per platform
+    pub fn add_locked_packages(mut self, locked_packages: LockedPackages) -> Self {
+        self.locked_packages
+            .insert(locked_packages.platform, locked_packages);
+        self
+    }
+
+    /// Build a conda_lock file
+    pub fn build(self) -> Result<CondaLock, CalculateContentHashError> {
+        let content_hash = self
+            .platforms
+            .iter()
+            .map(|plat| {
+                Ok((
+                    *plat,
+                    content_hash::calculate_content_hash(plat, &self.input_specs, &self.channels)?,
+                ))
+            })
+            .collect::<Result<HashMap<_, _>, CalculateContentHashError>>()?;
+
+        let lock = CondaLock {
+            metadata: LockMeta {
+                content_hash,
+                channels: self.channels,
+                platforms: self.platforms.iter().cloned().collect(),
+                sources: self.sources.unwrap_or_default(),
+                time_metadata: self.time_metadata,
+                git_metadata: self.git_metadata,
+                inputs_metadata: None,
+                custom_metadata: None,
+            },
+            package: self
+                .locked_packages
+                .into_values()
+                .flat_map(|package| package.build())
+                .collect(),
+            version: super::default_version(),
+        };
+        Ok(lock)
+    }
+}
+
+/// Shorthand for creating packages per platform
+struct LockedPackages {
+    pub locked_packages: Vec<LockedPackage>,
+    pub platform: Platform,
+}
+
+impl LockedPackages {
+    pub fn new(platform: Platform) -> Self {
+        Self {
+            locked_packages: Vec::new(),
+            platform,
+        }
+    }
+
+    /// Add a locked package
+    pub fn add_locked_package(mut self, locked_package: LockedPackage) -> Self {
```

Comment:
I always like it better to use `&mut self` in these kind of methods. You can still do proper chaining in that case but you can also do:

```rust
let mut locked_packages = ...
locked_packages.add_locked_package(...)
```

otherwise you always have to move the result back into the binding.

Reply:
I get that problem, but I kinda think that goes against the builder pattern anyways.

I guess I like this a bit more because otherwise it's hard to pass in the `LockedPackages` directly into the builder, as it will be a mutable reference. Also e.g Clap also does it in this way.